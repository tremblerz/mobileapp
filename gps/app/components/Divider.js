import styled from '@emotion/native';

/** A theme-aware horizontal divider that uses `theme.border` color */
export const Divider = styled.View`
  background-color: ${({ theme }) => theme.border};
  height: 1px;
  width: 100%;
`;

/** A theme-aware vertical divider that uses `theme.border` color */
export const VerticalDivider = styled.View`
  background-color: ${({ theme }) => theme.border};
  height: 100%;
  width: 1px;
`;
