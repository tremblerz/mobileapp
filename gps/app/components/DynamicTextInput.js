// Custome Text handling RTL and LTR text direction
import styled from '@emotion/native';

import languages from '../locales/languages';

// Themes to be added

const getCurrentDirection = () => languages.dir();

export const DynamicTextInput = styled.TextInput`
  writing-direction: ${getCurrentDirection};
`;
