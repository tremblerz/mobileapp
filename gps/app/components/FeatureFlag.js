import { useFlags } from '../helpers/Flags';

/**
 *
 * Usage:
 *
 * ```
 * <FeatureFlag
 *  name="google_import"
 *  fallback={<Text>Old</Text>}
 * >
 *   <FeatureUi />
 * </FeatureFlag>
 * ```
 *
 * @param {{
 *   name: string;
 *   fallback?: import('react').ReactNode;
 *   children: import('react').ReactNode;
 * }} param0
 */
export const FeatureFlag = ({ name, children, fallback }) => {
  const [flags] = useFlags();
  const fallbackRender = fallback || null;

  return flags[name] ? children : fallbackRender;
};
