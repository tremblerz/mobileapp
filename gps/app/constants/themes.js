import styled from '@emotion/native';
import { ThemeProvider } from 'emotion-theming';
import React from 'react';

import Color from './colors';

/** Violet on pale violet bg. e.g. Settings */
export const defaultTheme = {
  background: Color.VIOLET_ALPHA_06,
  textPrimaryOnBackground: Color.VIOLET,
  textSecondaryOnBackground: 'rgba(64, 81, 219, 0.6)',

  navBar: Color.VIOLET,
  onNavBar: Color.WHITE,

  /** E.g. button bg */
  primary: Color.VIOLET_BUTTON,
  /** E.g. button text color */
  onPrimary: Color.WHITE,

  success: Color.SUCCESS,
  warning: Color.WARNING,
  disabled: 'rgba(64, 81, 219, 0.6)',
  border: Color.DIVIDER,
};

/** White on violet bg. E.g. Main screen */
export const violet = {
  ...defaultTheme,
  background: `linear-gradient(245.21deg, ${Color.VIOLET_BUTTON} 0%, ${Color.VIOLET_BUTTON_DARK} 100%);`,
  textPrimaryOnBackground: Color.WHITE,
  textSecondaryOnBackground: 'rgba(255, 255, 255, 0.6)',

  navBarBorder: Color.VIOLET, // no visible border

  primary: Color.WHITE,
  onPrimary: Color.VIOLET,
  disabled: '#ececec',
};

/** White on gray bg. E.g. Possible exposure mode on default screen */
export const charcoal = {
  ...defaultTheme,
  background: Color.DARK_GRAY,
  textPrimaryOnBackground: Color.WHITE,
  textSecondaryOnBackground: Color.LIGHT_GRAY,

  navBar: Color.DARK_GRAY,
  navBarBorder: Color.DARK_GRAY, // no visible border

  primary: Color.WHITE,
  onPrimary: Color.DARK_GRAY,

  disabled: '#A9AFBA',
};

const THEME_MAP = {
  default: defaultTheme,
  violet,
  charcoal,
};

/** A view which uses the theme's background color */
export const ThemedBackground = styled.View`
  background-color: ${({ theme }) => theme.background};
`;

/**
 * Return theme config for a known theme name
 *
 * @param {'default' | 'violet' | 'charcoal'} themeName
 */
export const getTheme = themeName => {
  const themeColors = THEME_MAP[themeName];
  if (!themeColors) {
    throw new Error(`Unknown theme ${themeName}`);
  }
  return themeColors;
};

/**
 * Configures themeable sub-components with background, text (primary,
 * secondary), button and other various theme colors.
 *
 * Usage:
 *
 * This will render a charcoal view, with a white heading, and a grey paragraph.
 *
 * ```jsx
 * <Theme use="charcoal" setBackground>
 *   <Typography use="headline2">Test</Typogrphy>
 *   <Typography secondary>Paragraph here</Typogrphy>
 * </Theme>
 * ```
 *
 * It makes available the following `props.theme.x` properties in styled components:
 *
 * - background - the standard bg color
 * - primary - the primary theme/button color on top of bakground (e.g. violet)
 * - onPrimary - color to show on top of primary (usually white)
 * - onPrimary - color to show on top of primary (usually white)
 * - textPrimaryOnBackground - text color to show on background
 * - textSecondaryOnBackground - secondary text color to show on background
 * - warning - warning color (always orange)
 * - success - success color (always green)
 * - border - border/divider color (usually gray)
 *
 * @param {{
 *   use?: string,
 *   setBackground?: boolean,
 * }} param0
 */
export const Theme = ({ use = 'default', setBackground = false, children }) => {
  return (
    <ThemeProvider theme={getTheme(use)}>
      {setBackground ? (
        <ThemedBackground>{children}</ThemedBackground>
      ) : (
        children
      )}
    </ThemeProvider>
  );
};

/**
 * @param {{theme: defaultTheme, secondary?: boolean}} param0
 */
export const themeTextColor = ({ theme, secondary }) =>
  secondary ? theme.textSecondaryOnBackground : theme.textPrimaryOnBackground;
