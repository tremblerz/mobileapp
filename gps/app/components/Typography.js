import styled from '@emotion/native';
import { css } from '@emotion/native/dist/native.cjs.prod';
import * as React from 'react';

import { themeTextColor } from '../constants/themes';
import { useLanguageDirection } from '../locales/languages';

export const Type = {
  Headline1: 'headline1',
  Headline2: 'headline2',
  Headline3: 'headline3',
  Body1: 'body1',
  Body2: 'body2',
  Body3: 'body3',
};

/**
 * Render a theme and visual style aware text element.
 *
 * It uses the theme's `text(Primary|Secondary)OnBackground` color and a set of
 * predefined font size and line height values.
 *
 * Inspired by: https://material-components.github.io/material-components-web-catalog/#/component/typography
 *
 * Usage:
 *
 * ```jsx
 * <Theme use="charcoal">
 *   <Typography use="headline2">Heading</Typography>
 *   <Typography use="body1" secondary>Paragraph text ...</Typography>
 *   <Typography use="body2" monospace>link</Typography>
 * </Theme>
 * ```
 *
 * Use within a `<ThemeProvider>`
 *
 * @param {{
 *   use?: string,
 *   secondary?: boolean,
 *   monospace?: boolean,
 *   bold?: boolean,
 * }} param0
 */
export const Typography = ({
  use = Type.Body1,
  secondary,
  monospace,
  bold,
  style,
  children,
  ...otherProps
}) => {
  return (
    <ThemedText
      style={[
        style,
        css`
          writing-direction: ${useLanguageDirection()};
        `,
      ]}
      use={use}
      secondary={secondary}
      monospace={monospace}
      bold={bold}
      {...otherProps}>
      {children}
    </ThemedText>
  );
};

const FONT_SIZE_MAP = {
  [Type.Headline1]: '52px',
  [Type.Headline2]: '26px',
  [Type.Headline3]: '16px',
  [Type.Body1]: '18px',
  [Type.Body2]: '16px',
  [Type.Body3]: '15px',
};

const getFontSize = ({ use = Type.Body1 }) => FONT_SIZE_MAP[use];

const LINE_HEIGHT_MAP = {
  [Type.Headline1]: '52px',
  [Type.Headline2]: '34px',
  [Type.Headline3]: '40px',
  [Type.Body1]: '24px',
  [Type.Body2]: '22px',
  [Type.Body3]: '24px',
};

const getLineHeight = ({ use = Type.Body1 }) => LINE_HEIGHT_MAP[use];

const getTextOpacity = ({ disabled }) => (disabled ? '0.25' : null);

const getFontWeight = ({ use = Type.Body1, bold }) =>
  use.startsWith('headline') || bold ? 'bold' : 'normal';

const getFontFamily = ({ use, monospace, bold }) => {
  if (use.startsWith('headline') || bold) return 'IBMPlexSans-Bold';
  if (monospace) return 'IBMPlexMono';
  return 'IBMPlexSans';
};

const ThemedText = styled.Text`
  color: ${themeTextColor};
  opacity: ${getTextOpacity};
  font-family: ${getFontFamily};
  font-size: ${getFontSize};
  font-weight: ${getFontWeight};
  line-height: ${getLineHeight};
`;
