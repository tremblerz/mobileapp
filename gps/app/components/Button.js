import styled from '@emotion/native';
import { ThemeProvider } from 'emotion-theming';
import * as React from 'react';
import { SvgXml } from 'react-native-svg';

import { Typography } from './Typography';

/**
 * @param {{
 *   label: string;
 *   secondary?: boolean;
 *   icon?: string;
 *   onPress: () => void;
 *   disabled?: boolean;
 * }} param0
 */
export const Button = ({
  label,
  secondary,
  icon,
  onPress,
  disabled,
  small,
  ...otherProps
}) => {
  return (
    <Container
      onPress={onPress}
      accessible
      accessibilityLabel={label}
      accessibilityRole='button'
      hasIcon={!!icon}
      secondary={secondary}
      disabled={disabled}
      small={small}
      {...otherProps}>
      <ThemeProvider theme={invertTextColors}>
        <Label small={small} secondary={secondary} disabled={disabled}>
          {label}
        </Label>
        {icon ? <Icon xml={icon} /> : null}
      </ThemeProvider>
    </Container>
  );
};

const invertTextColors = theme => {
  return {
    ...theme,
    textPrimaryOnBackground: theme.onPrimary,
    textSecondaryOnBackground: theme.primary,
  };
};

const themePrimary = ({ theme, secondary, disabled }) =>
  disabled ? theme.disabled : secondary ? 'transparent' : theme.primary;

const getBorderColor = ({ theme, secondary }) =>
  secondary ? theme.primary : 'transparent';

const getJustifyContent = ({ hasIcon }) =>
  hasIcon ? 'space-between' : 'center';

const Container = styled.TouchableOpacity`
  background-color: ${themePrimary};
  height: ${({ small }) => (small ? '48px' : '72px')};
  border: 2px solid ${getBorderColor};
  padding-horizontal: ${({ small }) => (small ? '14px' : '16px')};
  border-radius: 8px;
  flex-direction: row;
  justify-content: ${getJustifyContent};
  align-content: center;
  align-items: center;
  align-self: stretch;
`;

const getFontSize = ({ small }) => (small ? '14px' : '20px');

const getLineHeight = ({ small }) => (small ? '21px' : '40px');

const Label = styled(Typography)`
  font-family: IBMPlexSans-Medium;
  font-size: ${getFontSize};
  line-height: ${getLineHeight};
  font-weight: normal;
`;

const Icon = styled(SvgXml)`
  margin-right: 8px; // for visual alignment
`;
