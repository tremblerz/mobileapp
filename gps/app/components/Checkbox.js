import React from 'react';
import { Image, TouchableOpacity } from 'react-native';

import boxCheckedIcon from '../../../shared/assets/images/boxCheckedIcon.png';
import boxUncheckedIcon from '../../../shared/assets/images/boxUncheckedIcon.png';
import { Typography } from './Typography';

export const Checkbox = ({ label, onPress, checked }) => {
  return (
    <TouchableOpacity
      style={{ flexDirection: 'row' }}
      onPress={onPress}
      accessible
      accessibilityLabel={label}>
      <Image
        source={checked === true ? boxCheckedIcon : boxUncheckedIcon}
        style={{ width: 25, height: 25, marginRight: 10 }}
      />
      <Typography use='body1'>{label}</Typography>
    </TouchableOpacity>
  );
};
