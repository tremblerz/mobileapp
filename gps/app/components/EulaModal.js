import React, { useEffect, useState } from 'react';
import { useTranslation } from 'react-i18next';
import { Linking, Modal, StyleSheet, View } from 'react-native';
import loadLocalResource from 'react-native-local-resource';
import { SafeAreaView } from 'react-native-safe-area-context';
import WebView from 'react-native-webview';

import close from '../../../shared/assets/svgs/close';
import Colors from '../constants/colors';
import { Theme } from '../constants/themes';
import en from '../locales/eula/en.html';
import ht from '../locales/eula/ht.html';
import { Button } from './Button';
import { Checkbox } from './Checkbox';
import { IconButton } from './IconButton';
import { Typography } from './Typography';

const EULA_FILES = { en, ht };

const DEFAULT_EULA_URL = 'about:blank';

export const EulaModal = ({ selectedLocale, continueFunction }) => {
  const [modalVisible, setModalVisibility] = useState(false);
  const [boxChecked, toggleCheckbox] = useState(false);
  const [html, setHtml] = useState(undefined);
  const { t } = useTranslation();

  // Pull the EULA in the correct language, with en as fallback
  const eulaPath = EULA_FILES[selectedLocale] || en;

  // Any links inside the EULA should launch a separate browser otherwise you can get stuck inside the app
  const shouldStartLoadWithRequestHandler = webViewState => {
    let shouldLoadRequest = true;
    if (webViewState.url !== DEFAULT_EULA_URL) {
      // If the webpage to load isn't the EULA, load it in a separate browser
      Linking.openURL(webViewState.url);
      // Don't load the page if its being handled in a separate browser
      shouldLoadRequest = false;
    }
    return shouldLoadRequest;
  };

  // Load the EULA from disk
  useEffect(() => {
    const loadEula = async () => {
      setHtml(await loadLocalResource(eulaPath));
    };
    loadEula();
  }, [selectedLocale, setHtml, eulaPath]);

  const canContinue = boxChecked;

  return (
    <>
      <Button
        label={t('label.launch_get_started')}
        onPress={() => setModalVisibility(true)}
      />
      <Modal animationType='slide' transparent visible={modalVisible}>
        <View style={styles.container}>
          <Theme use='default'>
            <SafeAreaView style={{ flex: 1 }}>
              <View style={{ flex: 7, paddingHorizontal: 5 }}>
                <IconButton
                  icon={close}
                  size={20}
                  style={styles.closeIcon}
                  accessibilityLabel='Close'
                  onPress={() => setModalVisibility(false)}
                />
                {html && (
                  <WebView
                    style={{ flex: 1 }}
                    source={{ html }}
                    onShouldStartLoadWithRequest={
                      shouldStartLoadWithRequestHandler
                    }
                  />
                )}
              </View>
            </SafeAreaView>
          </Theme>
          <Theme use='violet'>
            <SafeAreaView style={{ backgroundColor: Colors.VIOLET_BUTTON }}>
              <View style={styles.ctaBox}>
                <Checkbox
                  label={t('onboarding.eula_checkbox')}
                  onPress={() => toggleCheckbox(!boxChecked)}
                  checked={boxChecked}
                />
                <Typography style={styles.smallDescriptionText}>
                  {t('onboarding.eula_message')}
                </Typography>
                <Button
                  label={t('onboarding.eula_continue')}
                  disabled={!canContinue}
                  onPress={() => {
                    setModalVisibility(false);
                    continueFunction();
                  }}
                />
              </View>
            </SafeAreaView>
          </Theme>
        </View>
      </Modal>
    </>
  );
};

const styles = StyleSheet.create({
  // Container covers the entire screen
  container: {
    flex: 1,
    flexDirection: 'column',
    justifyContent: 'space-between',
    color: Colors.PRIMARY_TEXT,
    backgroundColor: Colors.WHITE,
  },
  ctaBox: {
    padding: 15,
    paddingTop: 0,
    backgroundColor: Colors.VIOLET_BUTTON,
  },
  closeIcon: {
    marginBottom: 6,
    marginRight: 8,
    alignSelf: 'flex-end',
  },
  smallDescriptionText: {
    fontSize: 14,
    marginVertical: 12,
  },
});
