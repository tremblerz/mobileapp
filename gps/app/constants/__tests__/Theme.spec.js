import { render } from '@testing-library/react-native';
import React from 'react';

import { Typography } from '../../components/Typography';
import { Theme } from '../themes';

it('includes extra background View if setBackground=true', () => {
  const { asJSON } = render(
    <Theme use='charcoal' setBackground>
      <Typography use='headline1'>Text</Typography>
    </Theme>,
  );

  expect(asJSON()).toMatchSnapshot();
});

it('changes text color based on theme', () => {
  const { asJSON } = render(
    <>
      <Theme use='charcoal'>
        <Typography use='body1'>Text</Typography>
      </Theme>
      <Theme use='default'>
        <Typography use='body1'>Text</Typography>
      </Theme>
    </>,
  );

  expect(asJSON()).toMatchSnapshot();
});
