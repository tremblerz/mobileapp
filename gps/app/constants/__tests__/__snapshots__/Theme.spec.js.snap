// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`changes text color based on theme 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Text
    style={
      Object {
        "color": "#FFF",
        "fontFamily": "IBMPlexSans",
        "fontSize": 18,
        "fontWeight": "normal",
        "lineHeight": 24,
        "writingDirection": "ltr",
      }
    }
    use="body1"
  >
    Text
  </Text>
  <Text
    style={
      Object {
        "color": "#4051DB",
        "fontFamily": "IBMPlexSans",
        "fontSize": 18,
        "fontWeight": "normal",
        "lineHeight": 24,
        "writingDirection": "ltr",
      }
    }
    use="body1"
  >
    Text
  </Text>
</View>
`;

exports[`includes extra background View if setBackground=true 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#3C475B",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#FFF",
          "fontFamily": "IBMPlexSans-Bold",
          "fontSize": 52,
          "fontWeight": "bold",
          "lineHeight": 52,
          "writingDirection": "ltr",
        }
      }
      use="headline1"
    >
      Text
    </Text>
  </View>
</View>
`;
